griptape/tools/google_drive/tool.py

```diff
@@ -0,0 +1,145 @@
+            return ListArtifact(items)
+
+        except Exception as e:
+            logging.error(e)
+            return ErrorArtifact(f"error retrieving files from Drive: {e}")
+
+    def _resolve_path_to_id(self, service, path: str) -> Optional[str]:
+        parts = path.split('/')
+        current_id = 'root'
+
+        for index, part in enumerate(parts):
+            if index == len(parts) - 1:  # If it's the last part of the path
+                query = f"name='{part}' and '{current_id}' in parents"
+            else:
+                query = f"name='{part}' and '{current_id}' in parents and mimeType='application/vnd.google-apps.folder'"
+
+            response = service.files().list(q=query).execute()
+            files = response.get('files', [])
+
+            if not files:
+                return None
+            current_id = files[0]['id']
+
+        return current_id
+
+    def upload_file(self, params: dict) -> InfoArtifact | ErrorArtifact:
+        from googleapiclient.http import MediaFileUpload
+
+        values = params['values']
+
+        try:
+            service = self._build_client(self.UPLOAD_FILE_SCOPES)
+            file_metadata = {'name': values['file_name']}
+            media = MediaFileUpload(values['file_path'], mimetype=values['mime_type'])
+            file = service.files().create(body=file_metadata, media_body=media, fields='id').execute()
+
+            return InfoArtifact(f'A file was successfully uploaded with ID {file.get("id")}.')
+
+        except Exception as e:
+            logging.error(e)
+            return ErrorArtifact(f"error uploading file to Drive: {e}")
+
+    def download_file(self, params: dict) -> BlobArtifact | ErrorArtifact:
+        values = params["values"]
+
+        try:
+            service = self._build_client(self.LIST_FILES_SCOPES)
+            file_id = self._resolve_path_to_id(service, values["file_path"])
+            if not file_id:
+                return ErrorArtifact(f"Could not find file: {values['file_path']}")
+
+            file_info = service.files().get(fileId=file_id).execute()
+            mime_type = file_info['mimeType']
+
+            google_export_mime_mapping = {
+                "application/vnd.google-apps.document":
+                    "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
+                "application/vnd.google-apps.spreadsheet":
+                    "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
+                "application/vnd.google-apps.presentation":
+                    "application/vnd.openxmlformats-officedocument.presentationml.presentation"
+            }
+
+            if mime_type in google_export_mime_mapping:
+                export_mime = google_export_mime_mapping[mime_type]
+                request = service.files().export_media(fileId=file_id, mimeType=export_mime)
+            else:
+                request = service.files().get_media(fileId=file_id)
```

Comment:
What's the difference between those two calls? Do we care about download vs export in the context of the LLM downloading it?

Reply:
So the if statement exports the Google docs as docx, Google sheets as xlsx and Google Slides as pptx. But the other type of files can just be downloaded as their own format.